File: autozig-math/zig/mat2.zig
const std = @import("std");

pub const Mat2 = extern struct {
    cols: [2][2]f32,

    pub const IDENTITY = Mat2{ .cols = .{
        .{ 1.0, 0.0 },
        .{ 0.0, 1.0 },
    } };

    pub const ZERO = Mat2{ .cols = .{
        .{ 0.0, 0.0 },
        .{ 0.0, 0.0 },
    } };

    pub fn identity() Mat2 {
        return IDENTITY;
    }

    pub fn from_cols(c0: Vec2, c1: Vec2) Mat2 {
        return .{ .cols = .{
            .{ c0.x, c0.y },
            .{ c1.x, c1.y },
        } };
    }

    pub fn from_angle(angle: f32) Mat2 {
        const c = @cos(angle);
        const s = @sin(angle);
        return .{ .cols = .{
            .{ c, s },
            .{ -s, c },
        } };
    }

    pub fn from_scale(scale: Vec2) Mat2 {
        return .{ .cols = .{
            .{ scale.x, 0.0 },
            .{ 0.0, scale.y },
        } };
    }

    pub fn mul(self: Mat2, other: Mat2) Mat2 {
        var result: Mat2 = undefined;
        inline for (0..2) |col| {
            inline for (0..2) |row| {
                var sum: f32 = 0.0;
                inline for (0..2) |k| {
                    sum += self.cols[k][row] * other.cols[col][k];
                }
                result.cols[col][row] = sum;
            }
        }
        return result;
    }

    pub fn mul_vec2(self: Mat2, v: Vec2) Vec2 {
        return .{
            .x = self.cols[0][0] * v.x + self.cols[1][0] * v.y,
            .y = self.cols[0][1] * v.x + self.cols[1][1] * v.y,
        };
    }

    pub fn transpose(self: Mat2) Mat2 {
        return .{ .cols = .{
            .{ self.cols[0][0], self.cols[1][0] },
            .{ self.cols[0][1], self.cols[1][1] },
        } };
    }

    pub fn determinant(self: Mat2) f32 {
        return self.cols[0][0] * self.cols[1][1] - self.cols[0][1] * self.cols[1][0];
    }

    pub fn inverse(self: Mat2) Mat2 {
        const det = self.determinant();
        if (@abs(det) < std.math.floatEps(f32)) {
            return IDENTITY;
        }
        const inv_det = 1.0 / det;
        return .{ .cols = .{
            .{ self.cols[1][1] * inv_det, -self.cols[0][1] * inv_det },
            .{ -self.cols[1][0] * inv_det, self.cols[0][0] * inv_det },
        } };
    }
};

export fn mat2_identity() Mat2 {
    return Mat2.identity();
}

export fn mat2_from_angle(angle: f32) Mat2 {
    return Mat2.from_angle(angle);
}

export fn mat2_from_scale(scale: Vec2) Mat2 {
    return Mat2.from_scale(scale);
}

export fn mat2_mul(self: Mat2, other: Mat2) Mat2 {
    return self.mul(other);
}

export fn mat2_mul_vec2(self: Mat2, v: Vec2) Vec2 {
    return self.mul_vec2(v);
}

export fn mat2_transpose(self: Mat2) Mat2 {
    return self.transpose();
}

export fn mat2_determinant(self: Mat2) f32 {
    return self.determinant();
}

export fn mat2_inverse(self: Mat2) Mat2 {
    return self.inverse();
}

File: autozig-math/src/mat2.rs
use autozig::include_zig;
use crate::Vec2;

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Mat2 {
    pub cols: [[f32; 2]; 2],
}

include_zig!("zig/mat2.zig", {
    fn mat2_identity() -> Mat2;
    fn mat2_from_angle(angle: f32) -> Mat2;
    fn mat2_from_scale(scale: Vec2) -> Mat2;
    fn mat2_mul(self_: Mat2, other: Mat2) -> Mat2;
    fn mat2_mul_vec2(self_: Mat2, v: Vec2) -> Vec2;
    fn mat2_transpose(self_: Mat2) -> Mat2;
    fn mat2_determinant(self_: Mat2) -> f32;
    fn mat2_inverse(self_: Mat2) -> Mat2;
});

impl Mat2 {
    pub const IDENTITY: Self = Self {
        cols: [[1.0, 0.0], [0.0, 1.0]],
    };

    pub const ZERO: Self = Self {
        cols: [[0.0, 0.0], [0.0, 0.0]],
    };

    pub fn identity() -> Self {
        mat2_identity()
    }

    pub fn from_angle(angle: f32) -> Self {
        mat2_from_angle(angle)
    }

    pub fn from_scale(scale: Vec2) -> Self {
        mat2_from_scale(scale)
    }

    pub fn mul_vec2(self, v: Vec2) -> Vec2 {
        mat2_mul_vec2(self, v)
    }

    pub fn transpose(self) -> Self {
        mat2_transpose(self)
    }

    pub fn determinant(self) -> f32 {
        mat2_determinant(self)
    }

    pub fn inverse(self) -> Self {
        mat2_inverse(self)
    }
}

impl std::ops::Mul for Mat2 {
    type Output = Self;
    fn mul(self, rhs: Self) -> Self::Output {
        mat2_mul(self, rhs)
    }
}

impl std::ops::Mul<Vec2> for Mat2 {
    type Output = Vec2;
    fn mul(self, rhs: Vec2) -> Self::Output {
        mat2_mul_vec2(self, rhs)
    }
}
